nvflare/fuel/hci/client/api.py

```diff
@@ -164,9 +170,56 @@ def _load_client_cmds(self, cmd_modules):
     def register_command(self, cmd_entry):
         self.all_cmds.append(cmd_entry.name)
 
+    def start_session_monitor(self, session_ended_callback, interval=5):
+        if self.sess_monitor_thread and self.sess_monitor_thread.is_alive():
+            self.close_session_monitor()
+        self.sess_monitor_thread = threading.Thread(
+            target=self._check_session, args=(session_ended_callback, interval), daemon=True
+        )
+        self.sess_monitor_active = True
+        self.sess_monitor_thread.start()
+
+    def close_session_monitor(self):
+        self.sess_monitor_active = False
+        if self.sess_monitor_thread:
+            self.sess_monitor_thread.join()
+
+    def _check_session(self, session_ended_callback, interval):
+        error_msg = ""
+        connection_error_counter = 0
+        while True:
+            time.sleep(interval)
+
+            if not self.sess_monitor_active:
+                return
+
+            if self.shutdown_received:
+                error_msg = self.shutdown_msg
+                break
+
+            resp = self.server_execute("_check_session")
+            status = resp["status"]
+
+            connection_error_counter += 1
+            if status != APIStatus.ERROR_SERVER_CONNECTION:
+                connection_error_counter = 0
+
+            if status in APIStatus.ERROR_INACTIVE_SESSION or (
+                status in APIStatus.ERROR_SERVER_CONNECTION and connection_error_counter > 60 // interval
+            ):
+                for item in resp["data"]:
+                    if item["type"] == "error":
+                        error_msg = item["data"]
+                break
+
+        self.server_sess_active = False
+        session_ended_callback(error_msg)
+
     def logout(self):
         """Send logout command to server."""
-        return self.server_execute("_logout")
+        resp = self.server_execute("_logout")
+        self.server_sess_active = False
```

Comment:
Since self.server_sess_active is set to False here, should the _check_session() method also check this flag? If it's False, then perhaps the monitor should end?

Reply:
If this flag gets set to False, then the next time _check_session() calls server_execute(), then it will receive APIStatus.ERROR_INACTIVE_SESSION and then the monitor will end.